### ISSUE TITLE
Provide capability for custom configuration

### ISSUE THREAD

**Issue by jgrandja** [spring-projects-experimental/spring-authorization-server#139]
The goal of this ticket is to capture all the possible configuration options that could require customization based on the requirements of an authorization server deployment.

**NOTE:** This ticket will be on hold until we reach a candidate MVP release, along with stability in the public API's.

The following is the _on-going_ list of possible configuration options:

- Provide plugin mechanism for extension authorization grant
- Configure a custom User Consent page

**Comment by Clem69B** [comment]
Hello ! 

An other proposition : 
> Provide plugin mechanism for Access Token customization (add custom claims in JWT token)

Thanks for this job !

**Comment by jgrandja** [comment]
@Clem69B JWT customization is already available via #173

**Comment by anoopgarlapati** [comment]
@jgrandja I have a couple more customization suggestions:
1. Provide ability for custom `redirect_uri` matching: Currently (release `0.1.0`) the `redirect_uri` from authorization request is matched strictly against the registered values from `RegisteredClient`. However, there is a provision in [OAuth 2.1 draft spec](https://tools.ietf.org/html/draft-ietf-oauth-v2-1-01#section-10.3.3) which allows any port to be specified for redirection on loopback address.
   > The authorization server MUST allow any port to be specified at the time of the request for loopback IP redirect URIs, to accommodate clients that obtain an available ephemeral port from the operating system at the time of the request.
   
   The above is more of a fix in the current in the current redirection matching than a new customization as we are attempting to conform to OAuth 2.1. 
   Also, there are use cases for authorization servers to relax exact string matching for `redirect_uri` in non-production or development environments. For this use case it would be better to allow custom matching of the redirect endpoint.
2. Provide ability to add custom token response parameters: Currently (release `0.1.0`) the token response is constructed from `OAuth2AccessTokenAuthenticationToken` which does have `additionalParameters` which would be written in the response. However, there is no convenient way to add parameters to this `Map` without completely overriding the `AuthenticationProvider` classes. Similar to customizing `Jwt` claims we should be able to add custom `additionalParameters` to the token.

**Comment by metacubed** [comment]
#### Provide a generic mechanism to reduce<sup>++</sup> the granted scopes present in a token

> Moving over #218 as recommended:

For either client- or user-based grants, the list of granted scopes is initially populated with those available to the client. This list is then filtered by other constraints, such as:

- The resource-owner consenting to only a subset of scopes
- The `scope` claim in the token request (#212)
- Custom policies defined in a larger authorization-server product (#187)

This can be modeled as a chain of `TokenCustomizer` implementations, which take in the initial token request and the initial granted list of scopes, and return a subset of those scopes at the end of the chain.

<br/>
<sup>++</sup> I think it is reasonable to only <i>reduce</i> the available scopes, not add to them. However, there might be use cases for both.

**Comment by jgrandja** [comment]
@anoopgarlapati 

> Provide ability to add custom token response parameters

Yes, this is needed. I have it in the list "Provide plugin mechanism for Customizing the Token Response".

FYI, `OAuth2AuthorizationServerConfigurer` needs to be enhanced in order to allow for easy configuration / customization for the `Filter`'s, `AuthenticationProvider`'s, etc. I'm planning this work early next month and hoping to get something in for next release.

This will be a critical piece in the overall configuration of the Authorization Server and I suspect it will take some time with a couple of iterations at least.

**Comment by braunsonm** [comment]
Great work on this project so far! I've been following it closely and am looking forward to using it. I'm wondering if there will be similar support as in spring security for providing an AuthenticationCustomizer. I would like to extend some of the authentication logic to authenticate a user based on the presence of a header and setup their security context based around that.

**Comment by jgrandja** [comment]
Thanks @braunsonm! Can you provide more specific details on the customization you're looking to perform so I can better understand.

**Comment by braunsonm** [comment]
Hey @jgrandja basically I would like to do custom authentication logic based on the presence of headers in the request. For instance if our LB terminates mTLS and sets a header to denote the user, I would have logic to read that header and setup the security context based on that username in the header. After that all the same logic of the authorization server should still work and issue a token
Hope that makes sense and might be possible!

**Comment by jgrandja** [comment]
@braunsonm Can you please log a new issue and provide as much detail as possible and we'll continue the discussion there. Thanks.

**Comment by wsaca** [comment]
My use case is related to #279, I'm trying to add **dynamically scopes** to the access token using information of the user who was logged in and the target audience passed as parameter. For now it is not possible to do it because if a client request a scope then it should be in **RegisteredClient** or the invalid scope exception is thown.

**Comment by jgrandja** [comment]
@wsaca The `OAuth2TokenCustomizer` gives you full control on enhancing the claims in the `Jwt` access token. It also provides you contextual information (e.g. the current user) that you may use to determine which claims to dynamically add/update/remove. Please see [comment](https://github.com/spring-projects-experimental/spring-authorization-server/pull/289#issuecomment-843992970).

**Comment by jgrandja** [comment]
@anoopgarlapati Instead of the framework defining a customization for a pre-defined default value for `scope`, I think this responsibility should be left to the custom `OAuth2TokenCustomizer` configured in the application. Thoughts?

**Comment by anoopgarlapati** [comment]
@jgrandja With the changes in #289, if the default behavior would be to reject the request when no scope is provided in authorize request, then it would not be possible for an application to customize default scopes using `OAuth2TokenCustomizer` as the execution never reaches there. 
My comment was to allow the restriction to be added in #289 as default behavior of the server (like Okta does for example) but also let that check be configurable so that using default scopes could be customized using `OAuth2TokenCustomizer` for those applications choosing that behavior. Does that make sense?

**Comment by jgrandja** [comment]
@anoopgarlapati Can you provide a link to the Okta docs so I can see how they implement.

**Comment by anoopgarlapati** [comment]
Hi @jgrandja You can find the [Okta developer documentation for authorize endpoint](https://developer.okta.com/docs/reference/api/oidc/#request-parameters) state that `scope` parameter is REQUIRED.
I have just tested it as well on my Okta dev server where I just provide `response_type`, `redirect_uri` and `client_id` without the `scope` parameter.
When I execute the request, I get `error=invalid_scope&error_description=No+scopes+were+requested.` error.

**Comment by wsaca** [comment]
@jgrandja In my example, if the requested scope was not added statically to "RegisteredClient" the client get an invalid scope error and OAuth2TokenCustomizer is never executed. A client will not always request scopes registered previously in a specific "RegisteredClient" instance because these are dynamically resolved, when #279 is implemented my case will be clearer.

**Comment by jgrandja** [comment]
@anoopgarlapati @wsaca Thank you for the detailed explanations.

@anoopgarlapati 

> allow the restriction to be added in #289 as default behaviour

Agreed.

> also let that check be configurable so that using default scopes could be customized using OAuth2TokenCustomizer

I think we could implement this as a policy where it's configured by default but may be customized. This would fulfill @wsaca use case, where this specific policy could be disabled in order to allow the execution to proceed to the `OAuth2TokenCustomizer` and scopes can be added there in a dynamic or default way.

How does this sound @anoopgarlapati @wsaca ?

**Comment by jgrandja** [comment]
@wsaca FYI, see [comment](https://github.com/spring-projects-experimental/spring-authorization-server/issues/288#issuecomment-857725029)
